pkg/config/spec.go

```diff
@@ -0,0 +1,147 @@
+		{"CPU", ""},
+		{"Processor", ""},
+		{"processor", ""},
+		{"@", ""},
+		{"®", ""},
+		{",", ""},
+	}
+
+	for _, r := range replacements {
+		name = strings.ReplaceAll(name, r.old, r.new)
+	}
+
+	// Remove "n-Bit Multi-Core" pattern
+	re := regexp.MustCompile(`\d+-Bit Multi-Core`)
+	name = re.ReplaceAllString(name, "")
+
+	// Remove anything in parentheses or square brackets
+	re = regexp.MustCompile(`[\(\[].*?[\)\]]`)
+	name = re.ReplaceAllString(name, "")
+
+	// Remove frequency pattern
+	re = regexp.MustCompile(`\d+(\.\d+)?\s?[GMgm][Hh]z`)
+	name = re.ReplaceAllString(name, "")
+
+	return strings.TrimSpace(name)
+}
+
+func formatProcessor(processor string) string {
+	name := rename(processor)
+	name = strings.ToLower(name)
+	// replace spaces and hyphens with an underscore
+	name = strings.ReplaceAll(name, "-", "_")
+	fields := strings.Fields(name)
+	name = strings.Join(fields, "_")
+	return strings.ReplaceAll(name, "_v", "v")
+}
+
+func formatVendor(vendor string) string {
+	vendor = strings.ToLower(vendor)
+	parts := strings.Fields(vendor)
+	if len(parts) > 0 {
+		vendor = parts[0]
+	}
+	vendor = strings.ReplaceAll(vendor, "-", "_")
+	vendor = strings.ReplaceAll(vendor, ",", "")
+	vendor = strings.ReplaceAll(vendor, "'", "")
+	return vendor
+}
+
+func roundToNearestHundred(value float64) int {
+	return int(math.Round(value/100) * 100)
+}
+
+func generateSpec() *MachineSpec {
+	spec := &MachineSpec{}
+	spec.Cores = -1
+	spec.Memory = -1
+	spec.Frequency = -1
+	spec.Chips = 1
+	spec.ThreadsPerCore = 1
+
+	cpus, err := cpu.Info()
+	if err == nil {
+		if len(cpus) > 0 {
+			spec.Processor = formatProcessor(cpus[0].ModelName)
+			spec.Vendor = formatVendor(cpus[0].VendorID)
+			spec.Frequency = roundToNearestHundred(cpus[0].Mhz) // rounded frequency
+		}
+	}
+
+	if cores, err := cpu.Counts(true); err == nil {
+		spec.Cores = cores
+	}
+	// Threads per core calculation
+	if coresPhysical, err := cpu.Counts(false); err == nil {
+		spec.ThreadsPerCore = spec.Cores / coresPhysical
+	}
+	// Memory info
+	if vmStat, err := mem.VirtualMemory(); err == nil {
+		spec.Memory = int(vmStat.Total / (1024 * 1024 * 1024)) // in GB
+	}
+	return spec
+}
+
+func readMachineSpec(path string) (MachineSpec, error) {
+	specFile, err := os.Open(path)
+	var spec MachineSpec
+	if err == nil {
+		if specBytes, err := io.ReadAll(specFile); err == nil {
+			if err = json.Unmarshal(specBytes, &spec); err == nil {
+				klog.Infof("Read machine spec from config: %v", spec)
+				return spec, nil
+			} else {
+				klog.Infof("failed to unmarshal machine spec %s: %v", string(specBytes), err)
+			}
+		}
+	}
+	return spec, err
+}
+
+// MachineSpec defines a machine spec to submit for power model selection
+type MachineSpec struct {
+	Vendor         string `json:"vendor"`
+	Processor      string `json:"processor"`
+	Cores          int    `json:"cores"`
+	Chips          int    `json:"chips"`
+	Memory         int    `json:"memory"`
+	Frequency      int    `json:"frequency"`
```

Comment:
along with CPU Freq, the `lscpu` commad also gives these fields
```
    CPU(s) scaling MHz:   19%
    CPU max MHz:          4800.0000
    CPU min MHz:          800.0000
```
Cpu Freq scaling has a direct impact on the power consumed by cpu, and we should consider this in power model. may be in another PR if not in this PR.



perhaps the go module does not give these fields.

Reply:
please include  `"CPU family"` and `"Model"` in the model 
```
Vendor ID:                GenuineIntel
  Model name:             11th Gen Intel(R) Core(TM) i7-11850H @ 2.50GHz
    CPU family:           6
    Model:                141
```

Reply:
Need update in both kepler-model-server. Need investigation and discussion later.